File: SecuredApp/Models/Product.swift
//
//  Product.swift
//  SecuredApp
//
//  Product model matching Supabase schema
//

import Foundation

enum ProductCondition: String, Codable, CaseIterable {
    case new = "new"
    case usedLikeNew = "used_like_new"
    case usedGood = "used_good"
    case usedFair = "used_fair"

    var displayName: String {
        switch self {
        case .new: return "New"
        case .usedLikeNew: return "Like New"
        case .usedGood: return "Good"
        case .usedFair: return "Fair"
        }
    }
}

struct Product: Identifiable, Codable, Hashable {
    let id: UUID
    let sku: String?
    let barcode: String?
    let name: String
    let description: String?
    let categoryId: UUID?

    // Product specifics
    let brand: String?
    let size: String?
    let condition: ProductCondition
    let colorway: String?
    let hasBox: Bool

    // Pricing
    let price: Decimal
    let cost: Decimal?
    let compareAtPrice: Decimal?

    // Inventory
    var quantity: Int
    let lowStockThreshold: Int

    // Images
    let images: [String]

    // Drops
    let isDrop: Bool
    let dropDate: Date?

    // External listings
    let ebayListingId: String?
    let whatnotListingId: String?

    // Metadata
    let isActive: Bool
    let isFeatured: Bool
    let tags: [String]

    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id, sku, barcode, name, description
        case categoryId = "category_id"
        case brand, size, condition, colorway
        case hasBox = "has_box"
        case price, cost
        case compareAtPrice = "compare_at_price"
        case quantity
        case lowStockThreshold = "low_stock_threshold"
        case images
        case isDrop = "is_drop"
        case dropDate = "drop_date"
        case ebayListingId = "ebay_listing_id"
        case whatnotListingId = "whatnot_listing_id"
        case isActive = "is_active"
        case isFeatured = "is_featured"
        case tags
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    // Computed properties
    var isInStock: Bool {
        quantity > 0
    }

    var isLowStock: Bool {
        quantity > 0 && quantity <= lowStockThreshold
    }

    var primaryImage: String? {
        images.first
    }

    var formattedPrice: String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "USD"
        return formatter.string(from: price as NSDecimalNumber) ?? "$0.00"
    }

    /// Returns true if this is a new drop (isDrop is true and created within last 7 days)
    var isNewDrop: Bool {
        guard isDrop else { return false }

        // Check if created within the last 7 days
        let sevenDaysAgo = Calendar.current.date(byAdding: .day, value: -7, to: Date()) ?? Date()
        return createdAt >= sevenDaysAgo
    }

    /// Formatted compare-at price for display
    var formattedCompareAtPrice: String? {
        guard let compareAt = compareAtPrice else { return nil }
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "USD"
        return formatter.string(from: compareAt as NSDecimalNumber)
    }

    /// Discount percentage from compare-at price
    var discountPercentage: Int? {
        guard let compareAt = compareAtPrice,
              compareAt > price else { return nil }
        let discount = ((compareAt - price) / compareAt) * 100
        return Int(truncating: discount as NSDecimalNumber)
    }

    /// Stock status for badge display
    var stockStatus: StockBadge.StockStatus {
        if quantity <= 0 {
            return .soldOut
        } else if isLowStock {
            return .lowStock
        } else {
            return .inStock
        }
    }
}

// Preview data
extension Product {
    static let preview = Product(
        id: UUID(),
        sku: "NK-AJ1-001",
        barcode: "123456789",
        name: "Air Jordan 1 Retro High OG",
        description: "The Air Jordan 1 Retro High OG brings back the classic silhouette.",
        categoryId: UUID(),
        brand: "Nike",
        size: "10",
        condition: .new,
        colorway: "Chicago",
        hasBox: true,
        price: 180.00,
        cost: 120.00,
        compareAtPrice: 200.00,
        quantity: 5,
        lowStockThreshold: 5,
        images: ["https://example.com/shoe1.jpg"],
        isDrop: false,
        dropDate: nil,
        ebayListingId: nil,
        whatnotListingId: nil,
        isActive: true,
        isFeatured: true,
        tags: ["jordan", "retro", "chicago"],
        createdAt: Date(),
        updatedAt: Date()
    )

    static let previewList: [Product] = [
        preview,
        Product(
            id: UUID(),
            sku: "NK-DUNK-002",
            barcode: "987654321",
            name: "Nike Dunk Low Panda",
            description: "Classic black and white colorway",
            categoryId: UUID(),
            brand: "Nike",
            size: "9.5",
            condition: .new,
            colorway: "Panda",
            hasBox: true,
            price: 110.00,
            cost: 70.00,
            compareAtPrice: nil,
            quantity: 12,
            lowStockThreshold: 5,
            images: [],
            isDrop: false,
            dropDate: nil,
            ebayListingId: nil,
            whatnotListingId: nil,
            isActive: true,
            isFeatured: false,
            tags: ["dunk", "panda"],
            createdAt: Date(),
            updatedAt: Date()
        )
    ]
}

File: website/src/types/product.ts
export type ProductCondition = "new" | "used_like_new" | "used_good" | "used_fair";

export const CONDITION_LABELS: Record<ProductCondition, string> = {
  new: "New",
  used_like_new: "Preowned",
  used_good: "Good",
  used_fair: "Fair",
};

// Sneaker-specific condition scale
export type VariantCondition = "DS" | "VNDS" | "Used-Excellent" | "Used-Good" | "Used-Fair";

export const VARIANT_CONDITION_LABELS: Record<VariantCondition, string> = {
  DS: "Deadstock (DS)",
  VNDS: "Very Near Deadstock",
  "Used-Excellent": "Used — Excellent",
  "Used-Good": "Used — Good",
  "Used-Fair": "Used — Fair",
};

export const VARIANT_CONDITIONS: VariantCondition[] = [
  "DS", "VNDS", "Used-Excellent", "Used-Good", "Used-Fair",
];

export interface ProductVariant {
  id: string;
  product_id: string;
  size: string | null;
  condition: VariantCondition;
  sku: string | null;
  barcode: string | null;
  price: number;
  cost: number | null;
  quantity: number;
  clover_item_id: string | null;
  created_at: string;
}

export interface Product {
  id: string;
  sku: string | null;
  barcode: string | null;
  name: string;
  description: string | null;
  category_id: string | null;
  brand: string | null;
  size: string | null;
  condition: ProductCondition;
  colorway: string | null;
  has_box: boolean;
  price: number;
  cost: number | null;
  compare_at_price: number | null;
  quantity: number;
  low_stock_threshold: number;
  images: string[];
  is_drop: boolean;
  drop_date: string | null;
  drop_price: number | null;
  drop_quantity: number | null;
  drop_starts_at: string | null;
  drop_ends_at: string | null;
  drop_sold_count: number;
  ebay_listing_id: string | null;
  whatnot_listing_id: string | null;
  is_active: boolean;
  is_featured: boolean;
  tags: string[];
  created_at: string;
  updated_at: string;
  inventory_location?: 'store' | 'warehouse';
  has_variants?: boolean;
  variants?: ProductVariant[];
}

export interface Category {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  image_url: string | null;
  sort_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

// Computed helpers
export function isInStock(product: Product): boolean {
  return product.quantity > 0;
}

export function isLowStock(product: Product): boolean {
  return product.quantity > 0 && product.quantity <= product.low_stock_threshold;
}

export function isNewDrop(product: Product): boolean {
  if (!product.is_drop) return false;
  const fiveDaysAgo = new Date();
  fiveDaysAgo.setDate(fiveDaysAgo.getDate() - 5);
  return new Date(product.created_at) >= fiveDaysAgo;
}

export type DropStatus = "upcoming" | "live" | "ended" | "sold_out";

export function getDropStatus(product: Product): DropStatus | null {
  if (!product.is_drop) return null;
  const now = new Date();
  const startsAt = product.drop_starts_at ? new Date(product.drop_starts_at) : null;
  const endsAt = product.drop_ends_at ? new Date(product.drop_ends_at) : null;
  const dropQty = product.drop_quantity;
  const soldCount = product.drop_sold_count ?? 0;

  if (dropQty !== null && dropQty !== undefined && soldCount >= dropQty) return "sold_out";
  if (endsAt && now >= endsAt) return "ended";
  if (startsAt && now < startsAt) return "upcoming";
  return "live";
}

export function isActiveDrop(product: Product): boolean {
  return getDropStatus(product) === "live";
}

export function getDropDisplayPrice(product: Product): number {
  if (product.is_drop && product.drop_price !== null && product.drop_price !== undefined) {
    return product.drop_price;
  }
  return product.price;
}

export function getDropRemainingQuantity(product: Product): number | null {
  if (!product.is_drop || product.drop_quantity === null || product.drop_quantity === undefined) return null;
  return Math.max(0, product.drop_quantity - (product.drop_sold_count ?? 0));
}

export function primaryImage(product: Product): string | null {
  return product.images?.[0] ?? null;
}

export function discountPercentage(product: Product): number | null {
  if (!product.compare_at_price || product.compare_at_price <= product.price) return null;
  return Math.round(
    ((product.compare_at_price - product.price) / product.compare_at_price) * 100
  );
}

export function formatCurrency(value: number): string {
  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
  }).format(value);
}

export type StockStatus = "in_stock" | "low_stock" | "sold_out";

export function getStockStatus(product: Product): StockStatus {
  if (product.quantity <= 0) return "sold_out";
  if (isLowStock(product)) return "low_stock";
  return "in_stock";
}
